### PR TITLE
feat(highlight): update default color scheme

### LINKED ISSUE #14790: Improved default color scheme
It was repeatedly brought up that the current default color scheme inherited from vim leaves something to be desired. So here is a proposal to improve the current default color scheme, where I am only interested in _objective_ improvements (so this is _not_ meant as a duplicate of "prettier default color schemes" or "bundle more modern color schemes", see https://github.com/neovim/neovim/issues/793 and https://github.com/vim/vim/issues/1665). For example:

1. Tweak the current colors (especially the diff colors) to be more legible.
2. In general, tweak the colors to be more accessible (for people with various forms of color vision deficiency).
3. Add more colors -- possibly as a separate truecolor variant, since the default is meant as a fallback that works reasonably well on every platform/terminal.
4. Add highlight groups for the new neovim-only features (LSP, tree-sitter).
5. Implement it in Lua.

The last point is of course a pure gimmick right now (but may still be nice for the sake of branding; at least currently, these also seem to bypass some minor open issues with vimscript color schemes) but may be more worthwhile if 0.6 includes a native Lua API for highlight groups (which, together with tree-sitter still being experimental for 0.5, is the reason for the milestone).

### PR DESCRIPTION
**Pull request by echasnovski**
**2023-12-16 Edit**: #26540 introduced some tweaks to the default color scheme based on the feedback. In particular, screenshots, palettes, and contrast ratios listed in this PR are now not up to date.

------

This PR addresses a long standing issue of suboptimal default color scheme. Here is the outline of what it proposes.

## Design

- Be "Neovim branded", i.e. follow [outlined example](https://github.com/neovim/neovim/issues/14790#issuecomment-927200052) as main reference from #14790. This generally means to mostly have "green-blue" feel plus one/two colors reserved for very occasional user attention.
- Be extra minimal for `notermguicolors` (256 colors) while allowing a bit more shades for when `termguicolors` is set (true colors). I **really** hope that this compromise will allow to coexist extreme minimalism of reference colors and a *bit* more variety with true colors.
- Be accessible, i.e. have enough [contrast ratio](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef). This means to have at least 7 for `Normal` highlight group and 4.5 for some common cases (`Visual`, `Comment` with set 'cursorline', colored syntax, `Diff*`, `Search`).
- Be dark and light variants a simple exchange of dark and light palettes (as this is easier to implement and requires less colors introduced).
- Be usable, i.e. provide enough visual feedback for common objects.

## Palette

- Introduce two separate palettes: dark (`NvimDark*`) and light (`NvimLight*`).
- Dark/light palette is used for background in dark/light color scheme and for foreground in light/dark color scheme. This introduces recognizable visual system without too standing out.
- Actual computation of palettes is done in a perceptually uniform [Oklch color space](https://bottosson.github.io/posts/oklab/). I highly recommend [this interactive picker/converter](https://bottosson.github.io/misc/colorpicker/).
- Each palette has the following colors (descriptions are for dark background; reverse for light one):
    - Four shades of "colored" greys for general UI. In 256 colors they are exact greys; with `termguicolors` - shades of "cool" grey.
        - Dark ones (from darkest to lightest) are reserved as background for `NormalFloat` (considered as "black"), `Normal` (background), `CursorLine`, `Visual`.
        - Light ones (also from darkest to lightest) are reserved for `Comment`, `StatusLine`/`TabLine`, `Normal` (foreground), and color considered as "white".
    - Hues for green, cyan, and red are modelled after reference color scheme: #87FFAF, #00E6E6, #D96D6A. Result have very similar hue but slightly different lightness and chroma (to be consistent with UI and palette).
    - Hues for yellow, blue, and magenta are chosen such as to be visibly different.
- Each palette color has the 256 colors variant with closest colors.

All colors are procedurally and objectively generated based on a handful of hyperparameters. Here is a [link to the script](https://gist.github.com/echasnovski/a04648c10d0b70e187e06068d2317e00) (to run, it requires ['mini.colors'](https://github.com/echasnovski/mini.colors) dependency).

Here is the data about palettes:

<details><summary>Hex palettes</summary>

```
Dark:
blue    = "#005078"
cyan    = "#007676"
green   = "#015825"
grey1   = "#0a0b10"
grey2   = "#1c1d23"
grey3   = "#2c2e33"
grey4   = "#4f5258"
magenta = "#4c0049"
red     = "#5e0009"
yellow  = "#6e5600"

Light:
blue    = "#9fd8ff"
cyan    = "#83efef"
green   = "#aaedb7"
grey1   = "#ebeef5"
grey2   = "#d7dae1"
grey3   = "#c4c6cd"
grey4   = "#9b9ea4"
magenta = "#ffc3fa"
red     = "#ffbcb5"
yellow  = "#f4d88c"
```

</details>

<details><summary>256 colors palettes</summary>

```
Dark:
blue    = 24
cyan    = 30
green   = 22
grey1   = 232
grey2   = 234
grey3   = 236
grey4   = 239
magenta = 53
red     = 52
yellow  = 58

Light:
blue    = 153
cyan    = 123
green   = 158
grey1   = 255
grey2   = 253
grey3   = 251
grey4   = 247
magenta = 189
red     = 217
yellow  = 222
```

</details>

<details><summary>Contrast ratios</summary>

`dark_*` are for dark color scheme, `ligth_*` - for light.

```
dark_blue         = 11
dark_comment      = 6.3
dark_comment_cur  = 5.1
dark_comment_vis  = 2.9
dark_cur_line     = 9.7
dark_cyan         = 12.5
dark_green        = 12.4
dark_magenta      = 11.6
dark_normal       = 12
dark_red          = 10.5
dark_visual       = 5.6
dark_yellow       = 12
light_blue        = 6.2
light_comment     = 5.6
light_comment_cur = 4.6
light_comment_vis = 2.9
light_cur_line    = 9.8
light_cyan        = 3.9
light_green       = 6.2
light_magenta     = 10.7
light_normal      = 12
light_red         = 10.1
light_visual      = 6.3
light_yellow      = 5
```

</details>

## Highlight groups

- Use greys for general UIs.
- Use bold text for keywords (`Statement` highlight group). This is an important choice to increase accessibility for people with color deficiencies, as it doesn't rely on actual color.
- Use green for strings, `DiffAdd` (as background), `DiagnosticOK`, and some minor text UI elements.
- Use cyan as main syntax color, i.e. for function calls (`Function`), `DiffText`, `DiagnosticInfo` (and `DiagnosticHint` in 256 colors), and some minor text UI elements.
- Use red to generally mean high user attention, i.e. errors. That is, `ErrorMsg`, `DiffDelete`, `DiagnosticError`.
- Use (**very sparingly**) yellow only with true colors to mean mild user attention, i.e. warnings. That is, `WarningMsg`, and `DiagnosticWarn`. With 256 colors it falls back to red. One important exception to this is `Search` which has yellow background.
- Use (**very sparingly**) blue as `Identifier` only with true colors. This gets used quite a lot, so it does not have fallback for 256 colors (to reduce noise).
- Magenta is not used and here only for completeness to allow setting full 8 terminal colors (later).

Here are screenshots in many situations:

<details><summary>Screenshots</summary>

Dark, 256 colors:

default-cs_dark_notgc

Dark, 256 colors, diff:

default-cs_dark_notgc_diff

Dark, 256 colors, tree-sitter:

default-cs_dark_notgc_treesitter

Dark, 256 colors, tree-sitter + LSP:

default-cs_dark_notgc_treesitter_lsp

Dark, true colors:

default-cs_dark_tgc

Dark, true colors, diff:

default-cs_dark_tgc_diff

Dark, true colors, tree-sitter:

default-cs_dark_tgc_treesitter

Dark, true colors, tree-sitter+LSP:

default-cs_dark_tgc_treesitter_lsp

Light, 256 colors:

default-cs_light_notgc

Light, 256 colors, diff:

default-cs_light_notgc_diff

Light, 256 colors, tree-sitter:

default-cs_light_notgc_treesitter

Light, 256 colors, tree-sitter + LSP:

default-cs_light_notgc_treesitter_lsp

Light, true colors:

default-cs_light_tgc

Light, true colors, diff:

default-cs_light_tgc_diff

Light, true colors, tree-sitter:

default-cs_light_tgc_treesitter

Light, true colors, tree-sitter+LSP:

default-cs_light_tgc_treesitter_lsp

</details>

------

@justinmk, as you are the one making merging decision, I am mentioning you here. To be fully upfront, here are the changes which I think you might not like (yet I still ask you to try them and reconsider):

- `Search` has yellow background. I tried various combinations, but having any other basic color (grey, green, cyan) as background *will* blend in more than it should. I tried `reverse=true`, but it looks not really good and has issue with not visible cursor (which is very often also reversed).
    - As yellow is not used anywhere in the syntax, having it as `Search` makes sense.
- There are small number of yellow (`Diagnostic*Warn`, `SpellCap`) and blue colors added with `termguicolors`. See screenshots.
- `Diff*` colors come from "background" palette. Those have similar lightness to background and stand out mostly via saturation (and not saturation + lightness). Reference examples in #14790 didn't have such colors available, so I decided to try them here. They do look good.

------

I would like this to be a present to Neovim for its 10-th anniversary (which is in two months, according to 72cf89bce8e4230dbc161dc5606f48ef9884ba70). Hope to have a discussion with eventual green light on actual color scheme being mergeable. After that I'll fix (probably literal) hundreds of failing tests (I patched some hanging ones, but couldn't see the finish of `make test` due to timeout).

------

Fixes https://github.com/neovim/neovim/issues/14790.
Resolves https://github.com/neovim/neovim/pull/24253.

### CODE REVIEW

**Review by gpanders — APPROVED** [review]

**Review by clason — APPROVED** [review]

**Review by seandewar — APPROVED** [review]

File: src/nvim/highlight_group.c
Comment (justinmk) [inline src/nvim/highlight_group.c:2887]: Defining our palette in the builtin color table seems like a great idea.

**Comment by justinmk** [comment]
Let's move quickly on this, @echasnovski you've been a heroic custodian of this topic, and it's a 2-way door--this is a good foundation for future improvements. Some minor requests:

- Most of your PR notes should be pasted into a new help file at `doc/dev_theme.txt`, so we can have guidelines as we develop the theme.
    - Also in the commit message!
- Hyperlinks should be underlined, as God intended (and github just noticed after 10 years 😂 ).
    - Probably also `:help` links, but can wait for now.

> Be accessible, i.e. have enough [contrast ratio](https://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef).

Yes. We should also consider bg that is closer black/white.

> `Search` has yellow background. I tried various combinations, but having any other basic color (grey, green, cyan) as background _will_ blend in more than it should.

Definitely not a blocker, though I'll mention that years ago I had yellow `Search` and eventually found it wasn't needed (the block background stands out pretty well, to start with).

> After that I'll fix (probably literal) hundreds of failing tests

We can ship the old Vim colors as a builtin `vim` theme. So we can force `:colorscheme vim` in the test runner, in order to unblock this first PR. Though I agree that we should probably update mosts tests (except where there's an extreme amount of churn--thoses tests could just do a local `:colorscheme vim` in their `before_each`) in a followup PR.

**Review by justinmk — APPROVED** [review]
💯 Let's do it

**Comment by echasnovski** [comment]
Thanks for approving!

>       * The "Highlight groups" notes especially are perfect.
>       * Also in the commit message!

Will do!
 
>     * Hyperlinks should be underlined, as God intended (and github just noticed after 10 years 😂 ).

Not sure what to do extra here. There seems to be only one group related to linking (`@text.uri`) and it was already underlined.

> We can ship the old Vim colors as a builtin `vim` theme. So we can force `:colorscheme vim` in the test runner, in order to unblock this first PR. Though I agree (in a followup PR) we should update mosts tests (except where there's an extreme amount of churn--then just do a local `:colorscheme vim` in `before_each`).

That certainly makes things easier! I am still committed to update all tests to make them proper if general core consensus decides so.

**Comment by glepnir** [comment]
dark:
-  `#005078`
-  `#007676`
-  `#015825`
- #0a0b10 `#0a0b10`
- #1c1d23 `#1c1d23`
- #2c2e33 `#2c2e33`
- #4f5258 `#4f5258`
- #4c0049 `#4c0049`
- #5e0009 `#5e0009`
- #6e5600 `#6e5600`

light:
- #9fd8ff `#9fd8ff`
- #83efef `#83efef`
- #aaedb7 `#aaedb7`
- #ebeef5 `#ebeef5`
- #d7dae1 `#d7dae1`
- #c4c6cd `#c4c6cd`
- #9b9ea4 `#9b9ea4`
- #ffc3fa `#ffc3fa`
- #ffbcb5 `#ffbcb5`
- #f4d88c `#f4d88c`

**Comment by clason** [comment]
Before other people waste their time on this: the palette is non-negotiable. Please only comment if you notice actual issues while tasting this theme.

**Review by lewis6991 — APPROVED** [review]

**Review by famiu — APPROVED** [review]

**Review by bfredl — APPROVED** [review]

**Comment by dundargoc** [comment]
This seems like a controversial feature. Maybe we should wait for more approvals before moving forward with this?

**Comment by clason** [comment]
y u no approve?? 😠

**Review by dundargoc — APPROVED** [review]

**Review by glepnir — APPROVED** [review]

**Comment by gpanders** [comment]
Don’t forget a news.txt entry.

**Comment by echasnovski** [comment]
> Don’t forget a news.txt entry.

And 'vim_diff.txt', yes.

**Comment by zeertzjq** [comment]
Also need `colorscheme vim` in `test/old/testdir/setup.vim`, after the `tlunmenu *`.

File: runtime/doc/vim_diff.txt
Comment (zeertzjq) [inline runtime/doc/vim_diff.txt:33]: I don't think "disabled" is the right word here.

File: test/functional/core/main_spec.lua
Comment (zeertzjq) [inline test/functional/core/main_spec.lua:83]:
```suggestion
        [1] = {bg_indexed = true, foreground = Screen.colors.Grey91, background = tonumber('0x161616'), fg_indexed = true},
        [2] = {bg_indexed = true, foreground = Screen.colors.Gray30, background = tonumber('0x161616'), fg_indexed = true},
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:28]:
```suggestion
Be suitable for dark and light backgrounds via a simple exchange of dark and light palettes.
```

(And bullet points?)

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:36]:
```suggestion
There are two separate palettes: dark and light. They all contain the same set of
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:41]:
```suggestion
The dark palette is used for background in tje dark color scheme and for
foreground in the light color scheme, and vice versa. This introduces recognizable visual
system without too standing out.
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:22]:
```suggestion
shades when 'termguicolors' is set (true colors).
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:18]:
```suggestion
Be "Neovim branded", i.e. have mostly "green-blue" feel plus one or two colors
```
Avoid ambiguous "either/or" slashes in favor of explicit connectives.

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:24]: What does "7" mean? Maybe add a link here to the definition.

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:50]:
```suggestion
  greys; in true colors they are shades of "cool" grey.
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:58]:
```suggestion
- Six colors to provide enough terminal colors: red, yellow, green,
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:61]:
```suggestion
  visually coherent. Lightness should be equal to the palette's basic grey (which
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:60]:
```suggestion
  They should have (reasonably) similar lightness and chroma to make them
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:65]:
```suggestion
- Each palette color should have a 256 colors variant with closest color
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:71]: Here (as in the design section), you could structure this as 

```
Use

* 8 shades of grey ...

* bold text ...

```

This makes the keywords stand out more.

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:84]:
```suggestion
Use red to generally mean high user attention, i.e. errors; in particular for
```

File: runtime/doc/dev_theme.txt
Comment (clason) [inline runtime/doc/dev_theme.txt:87]:
```suggestion
Use (very sparingly!) yellow only with true colors to mean mild user
```
Vimdoc is not Markup ;)

File: runtime/doc/news.txt
Comment (clason) [inline runtime/doc/news.txt:98]:
```suggestion
• Default color scheme has been updated to be "Neovim branded" and accessible. Use `:colorscheme vim` to revert to the old legacy color scheme.
```

File: runtime/doc/vim_diff.txt
Comment (clason) [inline runtime/doc/vim_diff.txt:33]:
```suggestion
- Default color scheme has been updated. This can be reversed by adding
```

**Comment by echasnovski** [comment]
@clason, @zeertzjq: All review notes are addressed. Thanks for the pointers!

File: src/nvim/highlight_group.c
Comment (tomtomjhj) [inline src/nvim/highlight_group.c:429]: `diffRemoved` links to `Special`, so removed lines in `diff` file type don't get highlighting. Not sure if that's worth handling in the default colorscheme, though.
image

Comment (clason) [reply src/nvim/highlight_group.c:429]: Yeah, the syntax files are kinda all over the place with the highlight groups -- this is one of the things we want to do better with treesitter, lessons learned. 

That being said, it may be worth contacting the syntax maintainers about links that can clearly be improved (like this one, which probably should link to DiffRemoved instead).

**Review by zeertzjq — APPROVED** [review]
LGTM. The two-digit ID numbers in `terminal/cursor_spec.lua` are a bit strange, but I guess this helps make future updates to `thelpers.screen_setup()`'s default ID numbers easier.

**Comment by clason** [comment]
Thank you, @echasnovski, for your hard work and patience in seeing this through! What a lovely (early) birthday present :)

File: runtime/colors/default.vim
Comment (justinmk) [reply runtime/colors/default.vim:5]: Should we mention the [`Nvim*` colors](https://github.com/neovim/neovim/blob/988b472d902e6116353fc2ae605bed32a37e6391/src/nvim/highlight_group.c#L2889-L2911) at `:help gui-colors` ?

**Comment by mrcjkb** [comment]
I've noticed some broken tree-sitter highlighting and git bisected it to this PR.

Specifically, Neovim used to highlight according to the last match in the highlights.scm file.
(I've been told this is the indended behaviour - and not to use priorities).

### Before this PR:

Notice how `spec` matches `@variable.haskell` last and is highlighted accordingly

treesitter


### After this PR:

Now `spec` is highlighted as `@function.haskell`

image

I'm not sure if this is known, and if there's already an issue for it (I couldn't find one). Since it's caused by this change, I thought I'd comment here before opening an issue :smile:

**Comment by clason** [comment]
Yes, treesitter groups have yet to be adapted; that is ongoing work (please read the description and open issues and PRs)

Friendly reminder: master is not guaranteed to be usable, if greater changes need to be carried out across multiple PRs.

File: src/nvim/highlight_group.c
Comment (unnamedname) [inline src/nvim/highlight_group.c:263]: Just updated to nvim 0.10, the biggest visual change is that all the variables have changed colors with a custom colorscheme, this strong contrast with the context seems to be more "visual overload" than before, I would suggest to keep linking to `Identifier`. Thanks!

Comment (clason) [reply src/nvim/highlight_group.c:263]: This is not up for debate; the colorscheme is final. Feel free to customize it yourself, of course.

**Pull request merged**
